go-abci-service/abci/monitor.go

```diff
@@ -42,37 +44,77 @@ func (app *AnchorApplication) SyncMonitor() {
 //Also ensures api is online
-func (app *AnchorApplication) KeyMonitor() {
+func (app *AnchorApplication) StakeIdentity() {
 	for app.JWKSent != true {
 		time.Sleep(60 * time.Second)
+		if _, exists := app.state.CoreKeys[app.ID]; exists {
+			return
+		}
 		if !app.state.ChainSynced {
 			continue
 		}
-		selfStatusURL := fmt.Sprintf("%s/status", app.config.APIURI)
-		response, err := http.Get(selfStatusURL)
+		amValidator, err := app.AmValidator()
 		if app.LogError(err) != nil {
 			continue
 		}
-		contents, err := ioutil.ReadAll(response.Body)
+		//if we're not a validator, we need to "stake" by opening a ln channel to the validators
+		if !amValidator {
+			validators, err := app.rpc.GetValidators(app.state.Height)
+			if app.LogError(err) != nil {
+				continue
+			}
+			for _, validator := range validators.Validators {
+				valID := validator.Address.String()
+				if lnID, exists := app.state.LnUris[valID]; exists {
+					app.logger.Info(fmt.Sprintf("Adding Lightning Peer %s...", lnID.Peer))
+					peerExists, err := app.lnClient.PeerExists(lnID.Peer)
+					app.LogError(err)
+					if peerExists || app.LogError(app.lnClient.AddPeer(lnID.Peer)) == nil {
+						chanExists, err := app.lnClient.ChannelExists(lnID.Peer, lnID.RequiredChanAmt)
+						app.LogError(err)
+						if !chanExists {
+							app.logger.Info(fmt.Sprintf("Adding Lightning Channel for Peer %s...", lnID.Peer))
+							_, err := app.lnClient.CreateChannel(lnID.Peer, lnID.RequiredChanAmt)
+							app.LogError(err)
+						} else {
+							app.logger.Info(fmt.Sprintf("Channel %s exists, skipping...", lnID.Peer))
+							continue
+						}
+					}
+				}
+			}
+		}
+		deadline := time.Now().Add(time.Duration(10*(app.lnClient.MinConfs+1)) * time.Minute)
+		for !time.Now().After(deadline) {
+			app.logger.Info("Sleeping to allow validator lightning channels to open...")
+			time.Sleep(time.Duration(1) * time.Minute)
+		}
+		jwk, err := jwk.New(app.config.ECPrivateKey.Public())
 		if app.LogError(err) != nil {
 			continue
 		}
-		var apiStatus types.CoreAPIStatus
-		err = json.Unmarshal(contents, &apiStatus)
+		jwkJson, err := json.MarshalIndent(jwk, "", "  ")
 		if app.LogError(err) != nil {
 			continue
 		}
-		app.JWK = apiStatus.Jwk
-		if app.JWK.Kid == "" {
+		//Create ln identity struct
+		resp, err := app.lnClient.GetInfo()
+		if app.LogError(err) != nil || len(resp.Uris) == 0 {
 			continue
 		}
-		jwkJson, err := json.Marshal(apiStatus.Jwk)
+		uri := resp.Uris[0]
+		lnID := types.LnIdentity{
+			Peer:            uri,
+			RequiredChanAmt: app.lnClient.LocalSats,
+		}
+		lnIDBytes, err := json.Marshal(lnID)
 		if app.LogError(err) != nil {
 			continue
 		}
-		_, err = app.rpc.BroadcastTx("JWK", string(jwkJson), 2, time.Now().Unix(), app.ID, &app.config.ECPrivateKey)
+		//Declare our identity to the network
+		_, err = app.rpc.BroadcastTxWithMeta("JWK", string(jwkJson), 2, time.Now().Unix(), app.ID, string(lnIDBytes), &app.config.ECPrivateKey)
```

Comment:
Is this what logs the node as a validator on the network? So after this call returns successfully then `AmValidator` would return true?

Reply:
Right. And the JWK transaction will only get allowed onto the chain by validators after it's passed the muster of VerifyIdentity()